bee-signing/src/ternary/mss.rs

```diff
@@ -0,0 +1,407 @@
+
+/// Errors occuring during MSS operations.
+#[derive(Debug, Error, PartialEq)]
+pub enum Error {
+    /// Invalid MSS depth provided.
+    #[error("Invalid MSS depth provided.")]
+    InvalidDepth(u8),
+    /// Missing MSS depth.
+    #[error("Missing MSS depth.")]
+    MissingDepth,
+    /// Missing signature index.
+    #[error("Missing signature index.")]
+    MissingSignatureIndex,
+    /// Missing underlying private key generator.
+    #[error("Missing underlying private key generator.")]
+    MissingGenerator,
+    /// Underlying private key generation failed.
+    #[error("Underlying private key generation failed.")]
+    FailedUnderlyingPrivateKeyGeneration,
+    /// Underlying public key generation failed.
+    #[error("Underlying public key generation failed.")]
+    FailedUnderlyingPublicKeyGeneration,
+    /// Underlying signature generation failed.
+    #[error("Underlying signature generation failed.")]
+    FailedUnderlyingSignatureGeneration,
+    /// Underlying public key recovery failed.
+    #[error("Underlying public key recovery failed.")]
+    FailedUnderlyingPublicKeyRecovery,
+    /// Public key generation failed.
+    #[error("Public key generation failed..")]
+    PublicKeyGenerationFailed,
+    /// Signature generation failed.
+    #[error("Signature generation failed..")]
+    SignatureGenerationFailed,
+    /// Failed sponge operation.
+    #[error("Failed sponge operation.")]
+    FailedSpongeOperation,
+    /// Seed generation failed.
+    #[error("Seed generation failed.")]
+    FailedSeedGeneration,
+    /// Signatures exhausted.
+    #[error("Signatures exhausted.")]
+    SignaturesExhausted,
+    /// Invalid signature size.
+    #[error("Invalid signature size.")]
+    InvalidSignatureSize,
+}
+
+/// Merkle Signature Scheme private key generator builder.
+pub struct MssPrivateKeyGeneratorBuilder<S, G> {
+    depth: Option<u8>,
+    generator: Option<G>,
+    sponge: PhantomData<S>,
+}
+
+impl<S, G> Default for MssPrivateKeyGeneratorBuilder<S, G>
+where
+    S: Sponge + Default,
+    G: PrivateKeyGenerator,
+{
+    fn default() -> Self {
+        Self {
+            depth: None,
+            generator: None,
+            sponge: PhantomData,
+        }
+    }
+}
+
+impl<S, G> MssPrivateKeyGeneratorBuilder<S, G>
+where
+    S: Sponge + Default,
+    G: PrivateKeyGenerator,
+{
+    /// Sets the depth of the MSS.
+    pub fn depth(mut self, depth: u8) -> Self {
+        self.depth.replace(depth);
+        self
+    }
+
+    /// Sets the underlying private key generator.
+    pub fn generator(mut self, generator: G) -> Self {
+        self.generator.replace(generator);
+        self
+    }
+
+    /// Builds the private key generator.
+    pub fn build(self) -> Result<MssPrivateKeyGenerator<S, G>, Error> {
+        let depth = match self.depth {
+            Some(depth) => match depth {
+                0..=MAX_MSS_DEPTH => depth,
+                _ => return Err(Error::InvalidDepth(depth)),
+            },
```

Comment:
I'ld check that in the setter already not even allowing an invalid state of the builder. The reason is *cautiousness*, because we cannot not know if some future dev decides to introduce a `build_xyz` method for some reason.

Reply:
Hmm I don't know. I feel like the whole point of having a builder is to possibly be in invalid state. Having only one method to check at the end seems easier. Also, I do not understand the second sentence.

Reply:
Never mind. Leave it as it is. My suggestion would make things more complicated, and another hypothetical `build_xyz` method would be a wrong implementation of the builder pattern.

Reply:
I guess an opinion of @zesterer and @christianpoveda on this would be helpful too. We should be consistent with all our builders.